s3/s3_operations.cpp

```diff
@@ -0,0 +1,1682 @@
+            msg << file_obj->physical_path();
+            msg << "\"";
+            if(data.status >= 0) {
+                msg << " - \"";
+                msg << S3_get_status_name((S3Status)data.status);
+                msg << "\"";
+            }
+            return ERROR(S3_FILE_UNLINK_ERR, msg.str());
+        }
+
+        return SUCCESS();
+
+    } // s3_file_unlink_operation
+
+
+    // =-=-=-=-=-=-=-
+    // interface for POSIX Stat
+    irods::error s3_file_stat_operation_with_flag_for_retry_on_not_found(
+        irods::plugin_context& _ctx,
+        struct stat* _statbuf,
+        bool retry_on_not_found )
+    {
+        unsigned long thread_id = std::hash<std::thread::id>{}(std::this_thread::get_id());
+        rodsLog(debug_log_level, "%s:%d (%s) [[%lu]]\n", __FILE__, __LINE__, __FUNCTION__, thread_id);
+
+        irods::error result = SUCCESS();
+
+        size_t retry_count_limit = S3_DEFAULT_RETRY_COUNT;
+        _ctx.prop_map().get<size_t>(s3_retry_count_size_t, retry_count_limit);
+
+        size_t retry_wait = S3_DEFAULT_RETRY_WAIT_SEC;
+        _ctx.prop_map().get<size_t>(s3_wait_time_sec_size_t, retry_wait);
+
+        // =-=-=-=-=-=-=-
+        // check incoming parameters
+        irods::error ret = s3CheckParams( _ctx );
+        if((result = ASSERT_PASS(ret, "[resource_name=%s] Invalid parameters or physical path.", get_resource_name(_ctx.prop_map()).c_str())).ok()) {
+
+            // =-=-=-=-=-=-=-
+            // get ref to fco
+            irods::data_object_ptr object = boost::dynamic_pointer_cast<irods::data_object>(_ctx.fco());
+
+            bzero (_statbuf, sizeof (struct stat));
+
+            boost::filesystem::path p(object->physical_path());
+            std::string filename = p.filename().string();
+
+            irods::error ret;
+            std::string bucket;
+            std::string key;
+            std::string key_id;
+            std::string access_key;
+
+            ret = parseS3Path(object->physical_path(), bucket, key, _ctx.prop_map());
+            if((result = ASSERT_PASS(ret, "[resource_name=%s] Failed parsing the S3 bucket and key from the physical path: \"%s\".", get_resource_name(_ctx.prop_map()).c_str(),
+                                     object->physical_path().c_str())).ok()) {
+
+                ret = s3InitPerOperation( _ctx.prop_map() );
+                if((result = ASSERT_PASS(ret, "[resource_name=%s] Failed to initialize the S3 system.", get_resource_name(_ctx.prop_map()).c_str())).ok()) {
+
+                    ret = s3GetAuthCredentials(_ctx.prop_map(), key_id, access_key);
+                    if((result = ASSERT_PASS(ret, "[resource_name=%s] Failed to get the S3 credentials properties.", get_resource_name(_ctx.prop_map()).c_str())).ok()) {
+
+                        std::string region_name = get_region_name(_ctx.prop_map());
+
+                        callback_data_t data;
+                        S3BucketContext bucketContext;// = {};
```

Comment:
why the bzero, rather than the {} initialization?    either way, let's remove the dead code.

Reply:
Removed the commented out code.

I am not sure why it needed the bzero but I believe that was required for some reason.